rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,408 @@
+The existing ways of doing this are:
+
+- Share nothing: Copy store objects to each consumer
+
+- Share everything: Use single mutable NFS store, carefully synchronizing updates.
+
+- Overlay everything: Mount all of `/nix` (store and DB) with OverlayFS.
+
+Each has serious drawbacks:
+
+- "Share nothing" wastes tons of space as many duplicate store objects are stored separately.
+
+- "Share everything" incurs major overhead from synchronization, even if consumers are making store objects they don't intend any other consumer to use.
+  It also poses an inflexible security model where the actions of one consumer effect all of them.
+
+- Overlay everything cannot take advantage of new store objects added to the lower store, because its "fork" of the DB covers up the lower store's.
+  (Furthermore, separate files from the DB proper like an out of date SQLite Write-Ahead-Logging (WAL) file *could* leak through, causing chaos.)
+
+The new `local-overlay` store also uses OverlayFS, but just for the store directory.
+The database is still a regular fresh empty one to start, and instead Nix explicitly knows about the lower store, so it can get any information for the DB it needs from it manually.
+This avoids all 3 downsides:
+
+- Store objects are never duplicated by the overlay store.
+  OverlayFS ensures that one copy in either layer is enough, and we are careful never to wastefully include a store object in the upper layer if it is already in the lower layer.
+
+- No excess synchronization.
+  Local changes are just that: local, not shared with any other consumer.
+  The lower store is never written to (no modifications or even filesystem locks) so we should not encounter any slow write and sync paths in filesystem implementations like NFS.
+
+- No rigidity of the lower store.
+  Since Nix sees both the `overlay-store`'s DB and the lower store, it is possible for it to be aware of and utilize objects that were added to the lower store after the upper store was created.
+
+This gives us a "best of all three worlds" solution.
+
+## Marketing motivation
+
+It is quite common for organizations using Nix to first adopt it behind the scenes.
+That is to say, Nix is used to prepare some artifacts which are then presented to a consumer that need not be aware they were made with Nix.
+Later though, because of Nix's gaining popularity, there may be a desire to reveal its usage so consumers can use Nix themselves.
+Rather than Nix being a controversial tool worth hiding, it can be a popular tool worth exposing.
+Nix-unware usage can still work, but Nix-aware usage can do additional things.
+
+The `local-overlay` store can serve as a crucial tool to bridge these two modes of using Nix.
+The lower store can be as before
+--- however the artifacts were disseminated in the "hidden Nix" first phase of adoption
+--- perhaps with only a small tweak to expose the DB / daemon socket if it wasn't before.
```

Comment:
Is exposing the socket used anywhere in the proposal, or is it just mentioned as a separate possibility (with relevant metadata sharing done via reading the underlying SQLite DB)?

Reply:
The local overlay store is potential a *consumer* of the socket provided by another Nix daemon. A Nix daemon can also be spun up using the local overlay store instead of the local store.

Basically, no new socket code is needed for this. As far as I can tell, everything one would want with sockets already works without limitation.

Reply:
A consumer — doing what with the socket?

Reply:
Doing regular client, like anything else using the daemon. It will in fact only use it to read metadata; the lower store can be read-only.

Reply:
I have reread the paragraph, and even with your explanations I am not sure what process the paragraph as written describes.

Reply:
1. Server Nix store dir, clients don't use Nix but do use Nix-built things
2. Serve either socket or SQLite database, clients can use Nix with that store but with restrictions (e.g. perhaps only read-only)
3. Use local overlay store (the writable upper layer makes the read-only lower layer less of an issue)

Doe that make sense?

Reply:
So in the step 1 each snapshot gets its own store path, and the rest of the store is visible? (I guess one could also share a fixed-path directory with links to all the currently-relevant snapshots?)